funsor/tensor.py

```diff
@@ -0,0 +1,973 @@
+        symbols = set(equation)
+        get_symbol = iter(map(opt_einsum.get_symbol, itertools.count()))
+        new_symbols = {}
+        for k in inputs:
+            symbol = next(get_symbol)
+            while symbol in symbols:
+                symbol = next(get_symbol)
+            symbols.add(symbol)
+            new_symbols[k] = symbol
+
+        # Manually broadcast using einsum symbols.
+        assert '.' not in equation
+        ins, out = equation.split('->')
+        ins = ins.split(',')
+        ins = [''.join(new_symbols[k] for k in x.inputs) + x_out
+               for x, x_out in zip(operands, ins)]
+        out = ''.join(new_symbols[k] for k in inputs) + out
+        equation = ','.join(ins) + '->' + out
+
+        data = ops.einsum(equation, *[x.data for x in operands])
+        return Tensor(data, inputs)
+
+    return None  # defer to default implementation
+
+
+def numeric_tensordot(x, y, dims):
+    """
+    Wrapper around :func:`torch.tensordot` or :func:`numpy.tensordot`
+    to operate on real-valued Funsors.
+
+    Note this operates only on the ``output`` tensor. To perform sum-product
+    contractions on named dimensions, instead use ``+`` and
+    :class:`~funsor.terms.Reduce`.
+
+    Arguments should satisfy::
+
+        len(x.shape) >= dims
+        len(y.shape) >= dims
+        dims == 0 or x.shape[-dims:] == y.shape[:dims]
+
+    :param Funsor x: A left hand argument.
+    :param Funsor y: A y hand argument.
+    :param int dims: The number of dimension of overlap of output shape.
+    :rtype: Funsor
+    """
+    assert dims >= 0
+    assert len(x.shape) >= dims
+    assert len(y.shape) >= dims
+    assert dims == 0 or x.shape[-dims:] == y.shape[:dims]
+    x_start, x_end = 0, len(x.output.shape)
+    y_start = x_end - dims
+    y_end = y_start + len(y.output.shape)
+    symbols = 'abcdefghijklmnopqrstuvwxyz'
+    equation = '{},{}->{}'.format(symbols[x_start:x_end],
+                                  symbols[y_start:y_end],
+                                  symbols[x_start:y_start] + symbols[x_end:y_end])
+    return Einsum(equation, (x, y))
+
+
+def numeric_stack(parts, dim=0):
+    """
+    Wrapper around :func:`torch.stack` to operate on real-valued Funsors.
+
+    Note this operates only on the ``output`` tensor. To stack funsors in a
+    new named dim, instead use :class:`~funsor.terms.Stack`.
+
+    :param tuple parts: A tuple of funsors.
+    :param int dim: A torch dim along which to stack.
+    :rtype: Funsor
+    """
+    assert isinstance(dim, int)
+    assert isinstance(parts, tuple)
+    assert len(set(x.output for x in parts)) == 1
+    shape = parts[0].output.shape
+    if dim >= 0:
+        dim = dim - len(shape) - 1
+    assert dim < 0
+    split = dim + len(shape) + 1
+    shape = shape[:split] + (len(parts),) + shape[split:]
+    output = Domain(shape, parts[0].dtype)
+    fn = functools.partial(ops.stack, dim)
+    return Function(fn, output, parts)
+
+
+REDUCE_OP_TO_NUMERIC = {
```

Comment:
Because we are making Tensor backend-agnostic, `torch.foo` is replaced by `ops.foo`. We already have `ops.min` and `ops.max`, so I used numpy [amin](https://docs.scipy.org/doc/numpy/reference/generated/numpy.amin.html), `amax` names here.